clinica/utils/dwi.py

```diff
@@ -1,253 +1,351 @@
-    data = im.get_fdata(dtype="float32")
-    hdr = im.get_header().copy()
-    bvals = np.loadtxt(in_bval)
-    bvecs = np.loadtxt(in_bvec)
+    if low_bval < 0:
+        raise ValueError(f"low_bval should be >=0. You provided {low_bval}.")
 
+    bvals = np.loadtxt(dwi_dataset.b_values)
+    bvecs = np.loadtxt(dwi_dataset.b_vectors)
     if bvals.shape[0] == bvecs.shape[0]:
         warnings.warn(
             "Warning: The b-vectors file should be column-wise. The b-vectors will be transposed",
             UserWarning,
         )
         bvecs = bvecs.T
 
-    lowbs = np.where(bvals <= low_bval)[0]
-
-    fname_b0, ext_b0 = os.path.splitext(os.path.basename(in_dwi))
-    if ext_b0 == ".gz":
-        fname_b0, ext2 = os.path.splitext(fname_b0)
-        ext_b0 = ext2 + ext_b0
-    out_b0 = os.path.abspath(f"{fname_b0}_b0{ext_b0}")
-    # out_b0 = op.abspath('b0.nii.gz')
-    b0data = data[..., lowbs]
-    hdr.set_data_shape(b0data.shape)
-    nib.Nifti1Image(b0data, im.get_affine(), hdr).to_filename(out_b0)
-
-    dwi_bvals = np.where(bvals > low_bval)[0]
-    out_dwi = os.path.abspath("dwi.nii.gz")
-    dwi_data = data[..., dwi_bvals]
-    hdr.set_data_shape(dwi_data.shape)
-    nib.Nifti1Image(dwi_data, im.get_affine(), hdr).to_filename(out_dwi)
-
-    bvals_dwi = bvals[dwi_bvals]
-    out_bvals = os.path.abspath("bvals")
-    np.savetxt(out_bvals, bvals_dwi, fmt="%d", delimiter=" ")
-
-    bvecs_dwi = np.array(
-        [
-            bvecs[0][dwi_bvals].tolist(),
-            bvecs[1][dwi_bvals].tolist(),
-            bvecs[2][dwi_bvals].tolist(),
-        ]
-    )
-    out_bvecs = os.path.abspath("bvecs")
-    np.savetxt(out_bvecs, bvecs_dwi, fmt="%10.5f", delimiter=" ")
-
-    return out_b0, out_dwi, out_bvals, out_bvecs
+    small_b_filter = get_b0_filter(dwi_dataset.b_values, low_bval=low_bval)
+    large_b_filter = np.array([i for i in range(len(bvals)) if i not in small_b_filter])
 
+    out_filename_small_dwi = add_suffix_to_filename(dwi_dataset.dwi, "small_b")
+    out_filename_large_dwi = add_suffix_to_filename(dwi_dataset.dwi, "large_b")
 
-def compute_average_b0(in_dwi: str, in_bval: str, low_bval: float = 5.0):
-    """Compute average b0 volume from DWI dataset."""
-    import os
-
-    import nibabel
-    import numpy as np
+    for volume_filter, out_filename in zip(
+        [small_b_filter, large_b_filter],
+        [out_filename_small_dwi, out_filename_large_dwi],
+    ):
+        data = compute_aggregated_volume(
+            dwi_dataset.dwi, aggregator=None, volumes_to_keep=volume_filter
+        )
+        img = get_new_image_like(dwi_dataset.dwi, data)
+        img.to_filename(out_filename)
 
-    assert os.path.isfile(in_dwi)
-    assert os.path.isfile(in_bval)
-    assert low_bval >= 0
+    out_filename_large_b_values = add_suffix_to_filename(
+        dwi_dataset.b_values, "large_b"
+    )
+    np.savetxt(
+        out_filename_large_b_values, bvals[large_b_filter], fmt="%d", delimiter=" "
+    )
```

Comment:
Since I was refactoring while you did this review, I'm not sure I have answered this comment or not. Let me know whether the current solution looks good from your perspective or not.

Reply:
yep it does !